nexus/db-queries/src/db/queries/region_allocation.rs

```diff
@@ -309,15 +309,40 @@ impl CandidateZpools {
             + diesel::dsl::sql(&zpool_size_delta.to_string()))
         .le(diesel::dsl::sql(zpool_dsl::total_size::NAME));
 
-        Self {
-            query: Box::new(
-                old_zpool_usage
-                    .query_source()
-                    .inner_join(with_zpool)
-                    .filter(it_will_fit)
-                    .select((old_zpool_usage::dsl::pool_id,)),
-            ),
-        }
+        let with_zpool = zpool_dsl::zpool
+            .on(zpool_dsl::id.eq(old_zpool_usage::dsl::pool_id));
+
+        let base_query = old_zpool_usage
+            .query_source()
+            .inner_join(with_zpool)
+            .filter(it_will_fit)
+            .select((old_zpool_usage::dsl::pool_id,));
+
+        let query = if distinct_sleds {
+            let seed_bytes = seed.to_le_bytes();
+
+            let query: Box<dyn CteQuery<SqlType = candidate_zpools::SqlType>> =
+                Box::new(
```

Comment:
(unrelated to this specific spot) I just wanna say, thanks again for going so deep into making this CTE magic work. This is a non-trivial query, and I like the way you've done it. Great work.